platform/run_image.go

```diff
@@ -38,65 +45,85 @@ func GetRunImageForExport(inputs LifecycleInputs) (RunImageForExport, error) {
-	if !image.Found() {
-		return &tm, nil
-	}
-	var err error
-	tm.OS, err = image.OS()
-	if err != nil {
-		return &tm, err
-	}
-	tm.Arch, err = image.Architecture()
-	if err != nil {
-		return &tm, err
-	}
-	tm.ArchVariant, err = image.Variant()
-	if err != nil {
-		return &tm, err
+func BestRunImageMirrorFor(targetRegistry string, runImageMD files.RunImageForExport, accessChecker ImageStrategy) (string, error) {
+	var runImageMirrors []string
+	if runImageMD.Image == "" {
+		return "", errors.New("missing run image metadata")
 	}
-	labels, err := image.Labels()
+	runImageMirrors = append(runImageMirrors, runImageMD.Image)
+	runImageMirrors = append(runImageMirrors, runImageMD.Mirrors...)
+
+	keychain, err := auth.DefaultKeychain(runImageMirrors...)
 	if err != nil {
-		return &tm, err
+		return "", fmt.Errorf("unable to create keychain: %w", err)
 	}
-	distName, distNameExists := labels[OSDistributionNameLabel]
-	distVersion, distVersionExists := labels[OSDistributionVersionLabel]
-	if distNameExists || distVersionExists {
-		tm.Distribution = &OSDistribution{Name: distName, Version: distVersion}
+
+	// Try to select run image on the same registry as the target
+	runImageRef := byRegistry(targetRegistry, runImageMirrors, accessChecker, keychain)
+	if runImageRef != "" {
+		return runImageRef, nil
 	}
-	if id, exists := labels[TargetLabel]; exists {
-		tm.ID = id
+
+	// Select the first run image we have access to
+	for _, image := range runImageMirrors {
+		ok, err := accessChecker.CheckReadAccess(image, keychain)
+		if err != nil {
+			return "", err
+		}
+		if ok {
+			return image, nil
+		}
 	}
 
-	return &tm, nil
+	return "", errors.New("failed to find accessible run image")
 }
 
-// Fulfills the prophecy set forth in https://github.com/buildpacks/rfcs/blob/b8abe33f2bdc58792acf0bd094dc4ce3c8a54dbb/text/0096-remove-stacks-mixins.md?plain=1#L97
-// by returning an array of "VARIABLE=value" strings suitable for inclusion in your environment or complete breakfast.
-func EnvVarsFor(tm *TargetMetadata) []string {
-	ret := []string{"CNB_TARGET_OS=" + tm.OS, "CNB_TARGET_ARCH=" + tm.Arch}
-	ret = append(ret, "CNB_TARGET_VARIANT="+tm.ArchVariant)
-	var distName, distVersion string
-	if tm.Distribution != nil {
-		distName = tm.Distribution.Name
-		distVersion = tm.Distribution.Version
+func byRegistry(reg string, repos []string, accessChecker ImageStrategy, keychain authn.Keychain) string {
+	for _, repo := range repos {
+		ref, err := name.ParseReference(repo, name.WeakValidation)
+		if err != nil {
+			continue
+		}
+		if reg == ref.Context().RegistryStr() {
+			ok, err := accessChecker.CheckReadAccess(repo, keychain)
+			if err != nil {
+				return ""
+			}
+			if ok {
+				return repo
+			}
+		}
 	}
-	ret = append(ret, "CNB_TARGET_DISTRO_NAME="+distName)
-	ret = append(ret, "CNB_TARGET_DISTRO_VERSION="+distVersion)
-	if tm.ID != "" {
-		ret = append(ret, "CNB_TARGET_ID="+tm.ID)
+	return ""
+}
+
+type ImageStrategy interface {
```

Comment:
imo we're not getting much mileage out of this interface:  because it's for exactly one function and the receivers are  stateless, it could be a "function signature type" and we could be passing function pointers as arguments.   that would reduce boilerplate and be more readable imo,  because you would have to make new empty structs just to access the function implementation. 

Do you think it's likely that we would add a second coupled function (so that the interface/strategy would be for a family of functions instead of just a single func?)

Reply:
SGTM. Updated in 67e81d09